src/libgrate/grate-asm.c

```diff
@@ -398,6 +398,46 @@ struct grate_shader *grate_shader_parse_fragment_asm(const char *asm_txt)
 		cgc->num_symbols++;
 	}
 
+	for (i = 0; i < 32 * 2; i++) {
+		if (!asm_fs_uniforms[i].used)
+			continue;
+
+		symbols = realloc(cgc->symbols,
+			(cgc->num_symbols + 1) * sizeof(struct cgc_symbol));
+
+		if (!symbols) {
+			free(cgc->symbols);
+			free(shader->cgc);
+			free(shader);
+			return NULL;
+		}
+
+		cgc->symbols = symbols;
+
+		cgc->symbols[cgc->num_symbols].location = i & 0x1F;
+		cgc->symbols[cgc->num_symbols].kind = GLSL_KIND_UNIFORM;
+		cgc->symbols[cgc->num_symbols].name = asm_fs_uniforms[i].name;
+		cgc->symbols[cgc->num_symbols].input = true;
+		cgc->symbols[cgc->num_symbols].used = true;
+
+		switch (asm_fs_uniforms[i].used) {
+		case FS_UNIFORM_FX10_LOW:
+			cgc->symbols[cgc->num_symbols].type = GRATE_TYPE_FX10_LOW;
+			break;
+		case FS_UNIFORM_FX10_HIGH:
+			cgc->symbols[cgc->num_symbols].type = GRATE_TYPE_FX10_HIGH;
+			break;
+		case FS_UNIFORM_FP20:
+			cgc->symbols[cgc->num_symbols].type = GRATE_TYPE_FP20;
```

Comment:
This seems like you're setting up a symobol per scalar value. Is that so? If so, I don't think it's a great idea.

I think it'd be better if we used the types already defined in glsl_type, and augmented it with a precision-value. We could use location to device between GRATE_TYPE_FX10_LOW and GRATE_TYPE_FX10_HIGH, by requiring GRATE_TYPE_FP20 to be aligned to multiples of two, and letting the last bit of location as a differentiator for FX10...